File: Trigger/TrigT1/TrigT1CaloMonitoring/src/GfexMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
*/

#include "GfexMonitorAlgorithm.h"
#include "CaloDetDescr/CaloDetDescrElement.h"
#include "CaloIdentifier/CaloGain.h"
#include "Identifier/Identifier.h"
#include "Identifier/HWIdentifier.h"
#include "CaloIdentifier/CaloCell_ID.h"
#include "CaloDetDescr/CaloDetDescrManager.h"
#include <istream>

GfexMonitorAlgorithm::GfexMonitorAlgorithm( const std::string& name, ISvcLocator* pSvcLocator )
  : AthMonitorAlgorithm(name,pSvcLocator)
{
}

StatusCode GfexMonitorAlgorithm::initialize() {
  ATH_MSG_DEBUG("GfexMonitorAlgorithm::initialize");
  ATH_MSG_DEBUG("Package Name "<< m_packageName);

  ATH_CHECK( m_gFexJetTobKeyList.initialize() ) ;
  ATH_CHECK( m_gFexRhoTobKeyList.initialize() ) ;
  ATH_CHECK( m_gFexGlobalTobKeyList.initialize() ) ;

  // Fill variable name map for global TOBs
  m_globTobVarMap.insert({"gScalarEJwoj", {"gFexMet", "gFexSumEt"}});
  m_globTobVarMap.insert({"gMETComponentsJwoj", {"METx", "METy"}});
  m_globTobVarMap.insert({"gMHTComponentsJwoj", {"MHTx", "MHTy"}});
  m_globTobVarMap.insert({"gMSTComponentsJwoj", {"MSTx", "MSTy"}});
  m_globTobVarMap.insert({"gMETComponentsNoiseCut", {"METx_NoiseCut", "METy_NoiseCut"}});
  m_globTobVarMap.insert({"gMETComponentsRms", {"METx_Rms", "METy_Rms"}});
  m_globTobVarMap.insert({"gScalarENoiseCut", {"gFexMet_NoiseCut", "gFexSumEt_NoiseCut"}});
  m_globTobVarMap.insert({"gScalarERms", {"gFexMet_Rms", "gFexSumEt_Rms"}});

  return AthMonitorAlgorithm::initialize();
}

StatusCode GfexMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {
  ATH_MSG_DEBUG("GfexMonitorAlgorithm::fillHistograms");

  // Small-R and large-R jets container loop
  for (const auto& key : m_gFexJetTobKeyList){
    SG::ReadHandle<xAOD::gFexJetRoIContainer> jetContainer (key, ctx);
    // Check that this container is present
    if ( !jetContainer.isValid() ) {
      ATH_MSG_WARNING("No gFex jet container found in storegate: "<< key.key());
    }
    else {
      const xAOD::gFexJetRoIContainer* jetContainerPtr = jetContainer.cptr();
      // Loop over all required pt cut values
      for(auto ptCut : m_ptCutValues){
        ATH_CHECK(fillJetHistograms(key.key(), jetContainerPtr, ptCut));
      }
    }
  } // end jet loop

  // Rho container loop
  for (const auto& key : m_gFexRhoTobKeyList){
    SG::ReadHandle<xAOD::gFexJetRoIContainer> rhoContainer (key, ctx);
    // Check that this container is present
    if ( !rhoContainer.isValid() ) {
      ATH_MSG_WARNING("No gFex rho container found in storegate: "<< key.key());
    }
    else {
      const xAOD::gFexJetRoIContainer* rhoContainerPtr = rhoContainer.cptr();
      ATH_CHECK(fillRhoHistograms(key.key(), rhoContainerPtr));
    }
  } // end rho container loop

  // Global TOB container loop
  for (const auto& key : m_gFexGlobalTobKeyList){
    SG::ReadHandle<xAOD::gFexGlobalRoIContainer> globalTobContainer (key, ctx);
    // Check that this container is present
    if ( !globalTobContainer.isValid() ) {
      ATH_MSG_WARNING("No gFex global TOB container found in storegate: "<< key.key());
    }
    else {
      const xAOD::gFexGlobalRoIContainer* globalTobContainerPtr = globalTobContainer.cptr();
      ATH_CHECK(fillGlobalTobHistograms(key.key(), globalTobContainerPtr));
    }
  } // end global TOBs container loop
  return StatusCode::SUCCESS;
}

StatusCode GfexMonitorAlgorithm::fillJetHistograms(const std::string& handleKey, const xAOD::gFexJetRoIContainer* container, const float& ptCutValue) const {

  // Define name extension based on pT cut value
  std::string histNameExt = ptCutValue != -1. ? (std::string("_CutPt") + std::to_string(int(ptCutValue))) : "";

  auto gtype = Monitored::Scalar<int>(handleKey + "gFexType", 0.0);
  auto jetEta = Monitored::Scalar<float>(handleKey + "Eta" + histNameExt, 0.0);
  auto jetPhi = Monitored::Scalar<float>(handleKey + "Phi" + histNameExt, 0.0);
  auto jetPt = Monitored::Scalar<float>(handleKey + "Pt" + histNameExt, 0.0);

  auto jetEtaFPGAa = Monitored::Scalar<float>(handleKey + "EtaFPGAa" + histNameExt, 0.0);
  auto jetEtaFPGAb = Monitored::Scalar<float>(handleKey + "EtaFPGAb" + histNameExt, 0.0);
  auto jetEtaFPGAc = Monitored::Scalar<float>(handleKey + "EtaFPGAc" + histNameExt, 0.0);

  auto jetPhiFPGAa = Monitored::Scalar<float>(handleKey + "PhiFPGAa" + histNameExt, 0.0);
  auto jetPhiFPGAb = Monitored::Scalar<float>(handleKey + "PhiFPGAb" + histNameExt, 0.0);
  auto jetPhiFPGAc = Monitored::Scalar<float>(handleKey + "PhiFPGAc" + histNameExt, 0.0);

  auto jetPtFPGAa = Monitored::Scalar<float>(handleKey + "PtFPGAa" + histNameExt, 0.0);
  auto jetPtFPGAb = Monitored::Scalar<float>(handleKey + "PtFPGAb" + histNameExt, 0.0);
  auto jetPtFPGAc = Monitored::Scalar<float>(handleKey + "PtFPGAc" + histNameExt, 0.0);

  for(const xAOD::gFexJetRoI* gFexJetRoI : *container){
    jetEta = gFexJetRoI->eta();
    jetPhi = gFexJetRoI->phi();
    jetPt  = gFexJetRoI->gFexTobEt();
    gtype  = gFexJetRoI->gFexType();

    if(jetPt > ptCutValue){
      fill(m_packageName, jetEta, jetPhi, jetPt, gtype);
      // Check for different pFPGAs
      FPGAType pFPGA(getFPGAType(jetEta));
      if (pFPGA == FPGAType::FPGAa){
        jetEtaFPGAa = gFexJetRoI->eta();
        jetPhiFPGAa = gFexJetRoI->phi();
        jetPtFPGAa  = gFexJetRoI->gFexTobEt();
        fill(m_packageName, jetEtaFPGAa, jetPhiFPGAa, jetPtFPGAa);
      }
      if (pFPGA == FPGAType::FPGAb){
        jetEtaFPGAb = gFexJetRoI->eta();
        jetPhiFPGAb = gFexJetRoI->phi();
        jetPtFPGAb  = gFexJetRoI->gFexTobEt();
        fill(m_packageName, jetEtaFPGAb, jetPhiFPGAb, jetPtFPGAb);
      }
      if (pFPGA == FPGAType::FPGAc){
        jetEtaFPGAc = gFexJetRoI->eta();
        jetPhiFPGAc = gFexJetRoI->phi();
        jetPtFPGAc  = gFexJetRoI->gFexTobEt();
        fill(m_packageName, jetEtaFPGAc, jetPhiFPGAc, jetPtFPGAc);
      }
    }
  }
  return StatusCode::SUCCESS;
}

StatusCode GfexMonitorAlgorithm::fillRhoHistograms(const std::string& handleKey, const xAOD::gFexJetRoIContainer* container) const {
  auto gFexRhoeT = Monitored::Scalar<float>(handleKey, 0.0);
  for(const xAOD::gFexJetRoI* gFexRhoRoI : *container){
      gFexRhoeT=gFexRhoRoI->gFexTobEt();
      fill(m_packageName, gFexRhoeT);
    }
  return StatusCode::SUCCESS;
}

StatusCode GfexMonitorAlgorithm::fillGlobalTobHistograms(const std::string& handleKey, const xAOD::gFexGlobalRoIContainer* container) const {
  // Find the variable names corresponding to the current key handle
  std::pair<std::string, std::string> varNames;
  for (const auto& [key, value] : m_globTobVarMap) {
    if (handleKey.find(key) != std::string::npos) {
        varNames = value;
        break;
    }
  }
  auto varOne = Monitored::Scalar<float>(varNames.first,0.0);
  auto varTwo = Monitored::Scalar<float>(varNames.second,0.0);

  for (const xAOD::gFexGlobalRoI* globRoI : *container) {
    varOne = globRoI->METquantityOne();
    varTwo = globRoI->METquantityTwo();
    fill(m_packageName, varOne, varTwo);
  }
  return StatusCode::SUCCESS;
}

GfexMonitorAlgorithm::FPGAType GfexMonitorAlgorithm::getFPGAType(const float& eta) const {
  if(eta < 0 && eta > -2.5) return FPGAType::FPGAa;
  if(eta > 0 && eta <  2.5) return FPGAType::FPGAb;
  if(std::abs(eta) > 2.5 && std::abs(eta)) return FPGAType::FPGAc;
  return FPGAType::None;
}

File: Trigger/TrigT1/TrigT1CaloMonitoring/python/GfexMonitorAlgorithm.py
#
#  Copyright (C) 2002-2023 CERN for the benefit of the ATLAS collaboration
#
def GfexMonitoringConfig(inputFlags):
    '''Function to configure LVL1 Gfex algorithm in the monitoring system.'''
    import math
    # get the component factory - used for merging the algorithm results
    from AthenaConfiguration.ComponentFactory import CompFactory
    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()
    #inputFlags.dump() # print all the configs

    # make the athena monitoring helper
    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(inputFlags,'GfexMonitoringCfg') 
    # get any algorithms
    GfexMonAlg = helper.addAlgorithm(CompFactory.GfexMonitorAlgorithm,'GfexMonAlg')

    # add any steering
    groupName = 'GfexMonitor' # the monitoring group name is also used for the package name
    GfexMonAlg.PackageName = groupName

    GfexMonAlg.gFexJetTobKeyList    = ["L1_gFexLRJetRoI", "L1_gFexSRJetRoI"]
    GfexMonAlg.gFexRhoTobKeyList    = ["L1_gFexRhoRoI"]
    GfexMonAlg.gFexGlobalTobKeyList = ["L1_gScalarEJwoj", "L1_gMETComponentsJwoj", "L1_gMHTComponentsJwoj", "L1_gMSTComponentsJwoj", "L1_gMETComponentsNoiseCut", "L1_gMETComponentsRms", "L1_gScalarENoiseCut", "L1_gScalarERms"]

    # Define various quantities
    mainDir = 'L1Calo'
    trigPath = 'Gfex/'
    globTobVarDict = {"gScalarEJwoj":["gFexMet", "gFexSumEt"], "gMETComponentsJwoj":["METx", "METy"], "gMHTComponentsJwoj":["MHTx", "MHTy"], "gMSTComponentsJwoj":["MSTx", "MSTy"], "gMETComponentsNoiseCut":["METx_NoiseCut", "METy_NoiseCut"], "gMETComponentsRms":["METx_Rms", "METy_Rms"], "gScalarENoiseCut":["gFexMet_NoiseCut", "gFexSumEt_NoiseCut"], "gScalarERms":["gFexMet_Rms", "gFexSumEt_Rms"]}
    keyDirPathMap  = {"gFexMet":"gXE/", "gFexSumEt":"gTE/", "METx":"gXE/", "METy":"gXE/", "MHTx":"gMHT/", "MHTy":"gMHT/", "MSTx":"gMST/", "MSTy":"gMST/", "METx_NoiseCut":"gXE_NoiseCut/", "METy_NoiseCut":"gXE_NoiseCut/", "METx_Rms":"gXE_RMS/", "METy_Rms":"gXE_RMS/", "gFexMet_NoiseCut":"gXE_NoiseCut/", "gFexSumEt_NoiseCut":"gTE_NoiseCut/", "gFexMet_Rms":"gTE_RMS/", "gFexSumEt_Rms":"gTE_RMS/"}
    globTobRangeDict = {"gFexMet":[0, 6e5], "gFexSumEt":[0, 5e6], "METx":[-3e5, 3e5], "METy":[-3e5, 3e5], "MHTx":[-3e5, 3e5], "MHTy":[-3e5, 3e5], "MSTx":[-2e5, 2e5], "MSTy":[-2e5, 2e5], "METx_NoiseCut":[-6e5, 6e5], "METy_NoiseCut":[-6e5, 6e5], "METx_Rms":[-6e5, 6e5], "METy_Rms":[-6e5, 6e5], "gFexMet_NoiseCut":[-1e4, 1e6], "gFexSumEt_NoiseCut":[0, 2.5e6], "gFexMet_Rms":[-1e5, 1e6], "gFexSumEt_Rms":[0, 2e6]}
    # add monitoring algorithm to group, with group name and main directory
    myGroup = helper.addGroup(GfexMonAlg, groupName , mainDir)

    # Define gfex histograms
    ptCutValues = GfexMonAlg.ptCutValues
    gFexJetTobKeyList = GfexMonAlg.gFexJetTobKeyList
    gFexRhoTobKeyList = GfexMonAlg.gFexRhoTobKeyList
    gFexGlobalTobKeyList = GfexMonAlg.gFexGlobalTobKeyList

    # Jet TOB list
    for containerKey in gFexJetTobKeyList:
        for ptCut in ptCutValues:
            for pFPGA in ["", "FPGAa", "FPGAb", "FPGAc"]:# Empty str for inclusive

                ptCutString = "_CutPt{:.0f}".format(ptCut) if (ptCut != -1) else ""
                containerKey = containerKey.split("+")[-1] # Needed to remove storeGate prefix if gFexJetTobKeyList is not set above

                # 1D histograms
                histKey = containerKey + "{}" + ptCutString
                histKey += ";h_" + containerKey + "{}" + ptCutString
                tobTypeStr = "gFex SRJet" if "SRJet" in containerKey else "gFex LRJet"
                ptStrTitle = " - tobEt [200 MeV Scale]>{:.0f}".format(ptCut) if ptCut != -1 else ""
                jPath = "gFexSRJets" if "SRJet" in containerKey else "gFexLRJets"
                jPath += ptCutString

                myGroup.defineHistogram(histKey.format("Eta{}".format(pFPGA), "Eta{}".format(pFPGA)), title="{} {} #eta{}; #eta; counts".format(tobTypeStr, pFPGA, ptStrTitle), type='TH1F', path=trigPath+"{}/".format(jPath), xbins=32,xmin=-3.3,xmax=3.3)
                myGroup.defineHistogram(histKey.format("Phi{}".format(pFPGA), "Phi{}".format(pFPGA)), title="{} {} #phi{}; #phi; counts".format(tobTypeStr, pFPGA, ptStrTitle), type='TH1F', path=trigPath+"{}/".format(jPath), xbins=32,xmin=-math.pi,xmax=math.pi)
                myGroup.defineHistogram(histKey.format("Pt{}".format(pFPGA),  "Pt{}".format(pFPGA)) , title="{} {} #Pt{} ; #Pt ; counts".format(tobTypeStr, pFPGA, ptStrTitle), type='TH1F', path=trigPath+"{}/".format(jPath), xbins=100,xmin=-1,xmax=4096)

                # 2D histograms
                histKey = containerKey + "{}" + ptCutString
                etaKey = histKey.format("Eta{}".format(pFPGA))
                phiKey = histKey.format("Phi{}".format(pFPGA))
                myGroup.defineHistogram('{},{};h_{}{}'.format(etaKey, phiKey, etaKey, phiKey), title="{} {} #eta vs #phi {}; #eta; #phi".format(tobTypeStr, pFPGA, ptStrTitle), type='TH2F',path=trigPath+"{}/".format(jPath), xbins=32,xmin=-3.3,xmax=3.3,ybins=32,ymin=-math.pi,ymax=math.pi)

    # Rho TOB list
    for containerKey in gFexRhoTobKeyList:
        myGroup.defineHistogram("{};h_{}".format(containerKey, containerKey), title="{}; gFexRho Et [MeV]; counts".format(containerKey), type="TH1F", path=trigPath+"gRHO", xbins=100,xmin=-300000,xmax=300000)

    # Global TOB list
    for containerKey in gFexGlobalTobKeyList:
        for key, dictVal in globTobVarDict.items():
            if key in containerKey:
                varOne, varTwo = dictVal
                break

        xminOne, xmaxOne = globTobRangeDict.get(varOne, [0,1e6])
        xminTwo, xmaxTwo = globTobRangeDict.get(varTwo, [0,1e6])

        myGroup.defineHistogram("{};h_{}".format(varOne, varOne), title="{}; {} [MeV]; counts".format(varOne, varOne), type="TH1F", path=trigPath+keyDirPathMap.get(varOne, "gFexGlob/"), xbins=100,xmin=xminOne,xmax=xmaxOne)
        myGroup.defineHistogram("{};h_{}".format(varTwo, varTwo), title="{}; {} [MeV]; counts".format(varTwo, varTwo), type="TH1F", path=trigPath+keyDirPathMap.get(varTwo, "gFexGlob/"), xbins=100,xmin=xminTwo,xmax=xmaxTwo)

    acc = helper.result()
    result.merge(acc)
    return result

if __name__=='__main__':
    # set input file and config options
    from AthenaConfiguration.AllConfigFlags import initConfigFlags
    import glob

    # Above MCs processed adding L1_eEMRoI
    inputs = glob.glob('/afs/cern.ch/user/t/thompson/work/public/LVL1_monbatch/run_sim/l1calo.361106.PowhegPythia8EvtGen_AZNLOCTEQ6L1_Zee.ESD.eFex_2021-05-16T2101.root')
    flags = initConfigFlags()
    flags.Input.Files = inputs
    flags.Output.HISTFileName = 'ExampleMonitorOutput.root'
    flags.lock()
    flags.dump() # print all the configs

    from AthenaCommon.AppMgr import ServiceMgr
    ServiceMgr.Dump = False
    from AthenaConfiguration.MainServicesConfig import MainServicesCfg  
    from AthenaPoolCnvSvc.PoolReadConfig import PoolReadCfg
    cfg = MainServicesCfg(flags)
    cfg.merge(PoolReadCfg(flags))

    GfexMonitorCfg = GfexMonitoringConfig(flags)
    cfg.merge(GfexMonitorCfg)
    cfg.printConfig(withDetails=False, summariseProps = True)

    nevents=100
    cfg.run(nevents)
